src/ng/q.js

```diff
@@ -223,22 +222,62 @@ function $QProvider() {
 
 function $$QProvider() {
   this.$get = ['$browser', '$exceptionHandler', function($browser, $exceptionHandler) {
-    return qFactory(function(callback) {
+    return qFactory(noop, function(callback, data) {
       $browser.defer(callback);
     }, $exceptionHandler);
   }];
 }
 
 /**
+ * @ngdoc service
+ * @name $qRaf
+ *
+ * @description
+ * `$qRaf` is a promise library that coordinates itself based on calls to `requestAnimationFrame`. The
+ * purpose of `$qRaf` is to allow animation-based code to be sequenced together without the need
+ * to hook into the digest cycle (which is how `$q` works).
+ *
+ * `$qRaf` is gauranteed to be delivered once one or more animation frames have passed. This means that
+ * even if the promise is resolved immediately then it will not be delivered until the next frame has passed.
+ * However if an animation does occur in between promise creation and resolution then the promise
+ * will be delivered immediately. This simple mechanism of flow control allows for animation callback
+ * code to avoid any unnecessary pauses. Any pause that occurs within an animation may lead to an
+ * unexpected reflow or page flicker and `$qRaf` aims to avoid this problem completely.
+ *
+ * Note that the API for `$qRaf` is the exact same as for `$q`.
+ *
+ * @param {function(function, function)} resolver Function which is responsible for resolving or
+ *   rejecting the newly created promise. The first parameter is a function which resolves the
+ *   promise, the second parameter is a function which rejects the promise.
+ *
+ * @returns {Promise} The newly created promise.
+ */
+function $QRafProvider() {
+  this.$get = ['$$rAF', '$exceptionHandler', function($$rAF, $exceptionHandler) {
+    return qFactory(
+      function(data) {
+        $$rAF(function() {
+          data.$$rafCleared = true;
+        });
+      },
+      function(callback, data) {
+        data.$$rafCleared ? callback() : $$rAF(callback);
```

Comment:
so you are calling the callback synchronously in some cases? that breaks all the guarantees that promises give you.